### PR TITLE
Sync the order of fields

### PR DESCRIPTION
**Pull request by rafalschmidt97**
Currently the solution doesn't check the order so if someone change in the existing field:

```csharp
[Display(Order = 30)]
```

to

```csharp
[Display(Order = 20)]
```

ContentSchemaSynchronizationService doens't see this as a change. Of course if you add a new fild with the order meta data it will be sorted to the right place but it doens't treat modification as a change.

For now I was forced to merge two branches here as there were some changes not on the master branch which were quite crutial for the solution but I'm gonna rebase after discussion this particular problem.

### CODE REVIEW

**Comment by rafalschmidt97** [comment]
@ArcusDeri could you elaborate why changes from `feature/non-content-base-classes` are not merged yet?

File: Forte.ContentfulSchema/Core/ContentTypeDefinition.cs [hunk 1]
```diff
@@ -80,6 +80,18 @@ public bool Update(ContentType contentType)
                     this.UpdateField(match.Existing, match.Updated, ref modified); 
                 }
             }
+
+            if (InferedContentType.Fields.Count == contentType.Fields.Count)
+            {
+                for (var i = 0; i < InferedContentType.Fields.Count; i++)
+                {
+                    if (InferedContentType.Fields[i]?.Id != contentType.Fields[i]?.Id) // order has changed
```
Comment (rafalschmidt97) [inline Forte.ContentfulSchema/Core/ContentTypeDefinition.cs]: @mgrzyb I feel like this is definitely enough to check the order. I've added to test cases and it works properly.

Whole solution still works:
image

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 2]
```diff
@@ -159,31 +131,69 @@ public void ShouldReturnFalseWhenUpdatingSameOrIncorrectEditorInterfaces(List<Ed
...
-//            var secondContentType = ContentTypeBuilder.New.WithFields(new Field {Id = "1"}).Build();
+        private static ContentType CreateDefaultContentType()
+        {
+            return new ContentType
+            {    
+                SystemProperties = new SystemProperties{Id = "123"},
+                Description = "Description",
+                DisplayField = "DisplayField",
+                Name = "Name",
+                Fields = new List<Field>{ new Field{ Id = "field1"} }
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:182]: Add one more fields so we can test changing order as one of the modification cases

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 3]
```diff
@@ -159,31 +131,69 @@ public void ShouldReturnFalseWhenUpdatingSameOrIncorrectEditorInterfaces(List<Ed
...
+            new Action<List<Field>>[] { f => f[0].Localized = true },
+            new Action<List<Field>>[] { f => f[0].Required = true },
+            new Action<List<Field>>[] { f => f[0].LinkType = "Entry" },
+            new Action<List<Field>>[] { f => f[0].Disabled = true },
+            new Action<List<Field>>[] { f => f[0].Validations = new List<IFieldValidator>(){new UniqueValidator()} },
+            new Action<List<Field>>[] { f => f[0].Items = new Schema(){LinkType = "TestLinkType"}},
+            new Action<List<Field>>[] { f => f[0].Items = new Schema(){Type = "TestType"}},
+            new Action<List<Field>>[] { f => f[0].Items = new Schema(){Validations = new List<IFieldValidator>(){new UniqueValidator()}}},
+            new Action<List<Field>>[] { f => f.Clear() },
+            new Action<List<Field>>[] { f => f.AddRange(new List<Field>{ new Field {Id = "NewField1"}, new Field {Id = "NewField2"} }) },
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:162]: Add new modification action that does: `f => f.Reverse()`

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 4]
```diff
@@ -159,31 +131,69 @@ public void ShouldReturnFalseWhenUpdatingSameOrIncorrectEditorInterfaces(List<Ed
...
 
-//        [Theory]
-//        [InlineData(false)]
-//        [InlineData(true)]
-//        public void ShouldReturnValueDependingFromTheFieldComparerResult(bool areEqual)
-//        {
-//            var fieldComparer = new Mock<IEqualityComparer<Field>>();
-//            fieldComparer.Setup(m => m.Equals(It.IsAny<Field>(), It.IsAny<Field>()))
-//                .Returns(areEqual);
+        public static readonly Action<List<Field>> OrderFieldsModificators = f =>
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:165]: This does not seem to be needed anymore (if the reversal of fields order is added as a modification action above.

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 5]
```diff
@@ -46,52 +43,27 @@ public void ShouldReturnTrueForUpdatedFields(Action<List<Field>> modifyFields)
...
+            Assert.True(definition.Update(originalContentType));
         }
 
-        private static ContentType CloneContentType(ContentType ct)
+        [Fact]
+        public void ShouldReturnFalseForNotUpdatedFieldsOrder()
         {
-            return JsonConvert.DeserializeObject<ContentType>(JsonConvert.SerializeObject(ct));
+            var originalContentType = CreateDefaultContentType();
+            OrderFieldsModificators(originalContentType.Fields);
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:62]: Remove

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 6]
```diff
@@ -46,52 +43,27 @@ public void ShouldReturnTrueForUpdatedFields(Action<List<Field>> modifyFields)
...
-                DisplayField = "DisplayField",
-                Name = "Name",
-                Fields = new List<Field>{ new Field{ Id = "field1"} }
-            };
+            Assert.True(definition.Update(originalContentType));
         }
 
-        private static ContentType CloneContentType(ContentType ct)
+        [Fact]
+        public void ShouldReturnFalseForNotUpdatedFieldsOrder()
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:59]: `ShouldReturnFalseWhenNoModifications`

Comment (mgrzyb) [reply Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:59]: Potentially remove if it duplicates `ShouldReturnFalseWhenUpdatingWithTheSameProperties`

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs [hunk 7]
```diff
@@ -46,52 +43,27 @@ public void ShouldReturnTrueForUpdatedFields(Action<List<Field>> modifyFields)
...
-        {
-            new Action<ContentType>[] {ct => ct.SystemProperties.Id = "321"},
-            new Action<ContentType>[] {ct => ct.Name = "NewName"},
-            new Action<ContentType>[] {ct => ct.Description = "NewDescription"},
-            new Action<ContentType>[] {ct => ct.DisplayField = "NewDisplayField"}
-        };
-
-        public static IEnumerable<object[]> FieldsModificators => new[]
+        [Fact]
+        public void ShouldReturnTrueForUpdatedFieldsOrder()
```
Comment (mgrzyb) [inline Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:47]: No longer needed. To be removed.

File: Forte.ContentfulSchema/Core/ContentTypeDefinition.cs [hunk 8]
```diff
@@ -74,8 +78,20 @@ public bool Update(ContentType contentType)
...
+                }
+            }
+
+            if (InferredContentType.Fields.Count == contentType.Fields.Count)
+            {
+                for (var i = 0; i < InferredContentType.Fields.Count; i++)
+                {
+                    if (InferredContentType.Fields[i]?.Id != contentType.Fields[i]?.Id) // order has changed
+                    {
+                            modified = true;
```
Comment (mgrzyb) [inline Forte.ContentfulSchema/Core/ContentTypeDefinition.cs:91]: When refactoring the tests, comment this out temporarily to make sure the tests still fail when order change is not detected.

File: Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs (hunk 3 above)
Comment (rafalschmidt97) [reply Forte.ContentfulSchema.Tests/Core/ContentTypeDefinitionTests.cs:162]: Great idea, thanks @mgrzyb :)

Two additional methods were unnecessary.

**Comment by rafalschmidt97** [comment]
> @ArcusDeri could you elaborate why changes from `feature/non-content-base-classes` are not merged yet?

Discussed with @mgrzyb.

**Pull request merged**